user guide: graph branching (#166)
cylc/cylc-doc: src/user-guide/writing-suites/configuration.rst

.. _User Guide Configuration:

Workflow Configuration
======================

Cylc suites are defined in structured, validated, :cylc:conf:`flow.cylc`
files that concisely specify the properties of, and the relationships
between, the various tasks managed by the suite.

> Here we will look at the format and content of the
> :cylc:conf:`flow.cylc` file and how to configure suites.

Reviewer: ```suggestion
Here we will look at:
- Folders which may (optionally) accompany a workflow configuration.
- The contents of the :cylc:conf:`flow.cylc` file.
- How to configure suites.
```

Author: This is just a file that I split into multiple parts, I didn't actually edit this but since you've gone to the effort to read it (and hats off for that) might as well take a look in.

.. _SuiteDefinitionDirectories:

> Suite Configuration Directories

Reviewer: ```suggestion
Workflow Configuration Directories
```

Author: I've purposfully not made this change, will need to make a full pass at the docs and code rather than patchily attempting as we go.

Also terminology isn't quite settled yet, is a Cylc workflow a "flow", but when you "re-flow" a workflow you get a "flow" with a "flow-label", is that a "flow" of the "workflow" if so what is a "run" of a "workflow", wait I know, it's a "flow" 🤦.

-------------------------------

> A Cylc :term:`suite directory` contains:

Reviewer: ```suggestion
A Cylc :term:`workflow directory` contains:
```
Don't worry if this is going to spread ripples all over cylc-doc. All our codebases and brains will require a `sed 's/suites/workflows/'` at some point.

:cylc:conf:`flow.cylc`
   The file which configures the workflow.
``bin/`` (optional)
   A directory where you can put scripts and executables for use
   in the workflow. It is automatically added to ``$PATH`` in the job
   execution environment.

   Alternatively, tasks can call external commands; or they can be
   scripted entirely within the flow.cylc file.
``lib/python/`` (optional)
   A directory for Python modules which can be used for:

   - Tasks; this directory is automatically added to ``$PYTHONPATH``
     in the job execution environment.
   - Custom :ref:`job submission modules <CustomJobSubmissionMethods>`.
   - Custom :ref:`Jinja2 Filters<CustomJinja2Filters>`).

> Other files and folders may be placed in the :term:`suite directory` e.g.

Reviewer: suite -> workflow.

documentation, configuration files, etc.

.. _FlowConfigFile:

flow.cylc File Overview
-----------------------

> :cylc:conf:`flow.cylc` files are an extended-INI format with section nesting.

Reviewer: Hostage to fortune putting external links in, but is it worth putting in a link to somewhere on what an INI format is? It's pretty long in the tooth.

organize large suites. Inclusion boundaries are completely arbitrary -
you can think of include-files as chunks of the :cylc:conf:`flow.cylc` file simply
cut-and-pasted into another file. Include-files may be included
multiple times in the same file, and even nested. Include-file paths
can be specified portably relative to the suite configuration directory,
e.g.:

.. code-block:: cylc

   # include the file $CYLC_SUITE_DEF_PATH/inc/foo.cylc:
   %include inc/foo.cylc

.. note::

   Template processors may have their own include functionality
   which can also be used.

.. note::

   Cylc's native file inclusion mechanism supports optional inlined
   editing:

   .. code-block:: bash

      $ cylc edit --inline SUITE

   The suite will be split back into its constituent include-files when you
   exit the edit session. While editing, the inlined file becomes the
   official suite configuration so that changes take effect whenever you save
   the file. See ``cylc prep edit --help`` for more information.

.. _SyntaxHighlighting:

Syntax Highlighting For Suite Configuration
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Cylc provides syntax plugins for the following editors:

> .. _Cylc.tmbundle: https://github.com/cylc/Cylc.tmbundle

Reviewer: Are we just not advertising the gedit, kate, vi &c ones any more?

Gedit
   The syntax file can be obtained from the Cylc library by
   running the following command
   ``cylc extract-resources . etc/syntax/cylc.lang``
   installation instructions are at the top of the file.
Kate
   The syntax file can be obtained from the Cylc library by
   running the following command
   ``cylc extract-resources . etc/syntax/cylc.xml``
   installation instructions are at the top of the file.
PyCharm
   Install the `Cylc.tmbundle`_.
Vim
   The syntax file can be obtained from the Cylc library by
   running the following command
   ``cylc extract-resources . etc/syntax/cylc.vim``
   installation instructions are at the top of the file.
Visual Studio Code
   Install the `vscode-cylc`_ extension.
Sublime Text 3
   Install the `Cylc.tmbundle`_.
TextMate
   Install the `Cylc.tmbundle`_.
WebStorm
      Install the `Cylc.tmbundle`_.

Gross File Structure
^^^^^^^^^^^^^^^^^^^^

Cylc :cylc:conf:`flow.cylc` files consist of a suite title and description followed by
configuration items grouped under several top level section headings:

:cylc:conf:`[cylc]`
   Non task-specific suite configuration.
:cylc:conf:`[scheduling]`
   Determines when tasks are ready to run.

>    - tasks with special behaviour, e.g. clock-trigger tasks

Reviewer: Worth a link to clock-trigger tasks for for the reader who asks WTF?

   - the dependency graph, which defines the relationships
     between tasks
:cylc:conf:`[runtime]`
   Determines how, where, and what to execute when tasks are ready

   - script, environment, job submission, remote hosting, etc.
   - suite-wide defaults in the *root* namespace
   - a nested family hierarchy with common properties
     inherited by related tasks
:cylc:conf:`[visualization]`
   Suite graph styling

.. _Validation:

Validation
^^^^^^^^^^

The ``cylc validate`` command evaluates the :cylc:conf:`flow.cylc` file
against a specification that defines all legal entries, values and options.
It also performs some integrity checks designed to catch certain configuration
issues and impossible scheduling constraints.

These checks are also performed by ``cylc run`` before starting a workflow.

> All legal entries are documented in :cylc:conf:`flow.cylc`.

Reviewer: Could you link straight to the reference page?
